js/packages/teams-ai/src/authentication/MessagingExtensionAuthentication.ts

```diff
@@ -0,0 +1,111 @@
+import { ActivityTypes, InvokeResponse, TokenResponse, TurnContext } from 'botbuilder';
+import { FETCH_TASK_INVOKE_NAME, QUERY_INVOKE_NAME, QUERY_LINK_INVOKE_NAME } from '../MessageExtensions';
+import * as UserTokenAccess from './UserTokenAccess';
+import { OAuthPromptSettings } from 'botbuilder-dialogs';
+
+/**
+ * @internal
+ */
+export class MessagingExtensionAuthentication {
+    public async authenticate(context: TurnContext, settings: OAuthPromptSettings): Promise<string | undefined> {
+        const authObj = context.activity.value.authentication;
+
+        // Token Exchange
+        if (authObj && authObj.token) {
+            // Message extension token exchange invoke activity
+            const isTokenExchangable = await this.isTokenExchangeable(context, settings);
+            if (!isTokenExchangable) {
+                await context.sendActivity({
+                    value: { status: 412 } as InvokeResponse,
+                    type: ActivityTypes.InvokeResponse
+                });
+
+                return undefined;
+            } else {
+                const tokenExchangeResponse = await this.exchangeToken(context, settings);
+                if (tokenExchangeResponse && tokenExchangeResponse.token) {
+                    return tokenExchangeResponse.token;
+                }
+            }
+        }
+
+        const value = context.activity.value;
+
+        // When the Bot Service Auth flow completes, the query.State will contain a magic code used for verification.
+        const magicCode = value.state && Number.isInteger(Number(value.state)) ? value.state : '';
+
+        const tokenResponse = await UserTokenAccess.getUserToken(context, settings, magicCode);
+
+        if (this.isValidActivity(context) && (!tokenResponse || !tokenResponse.token)) {
```

Comment:
minor: since the logic will only be called when `Authentication` class found it's a valid Messaging Extension activity, do we still need to do the check here?

Reply:
Yeah, for instance let's say the incoming activity is `composeExtension/submitAction`. If there's a token in token store, then this method will return the token. But if it doesn't then it will try to trigger the sign in flow, which is undesirable. And one could argue, makes the `MessagingExtension.authenticate` method buggy.

You can argue that we check if the user is logged in in the `Authenticate.signUserIn` method which is redundant and can be removed. But I put it in for readability and clarity sakes.